arangod/RocksDBEngine/RocksDBMetaCollection.cpp

```diff
@@ -424,114 +431,283 @@ std::unique_ptr<containers::RevisionTree> RocksDBMetaCollection::computeRevision
-  if (applied > _revisionTreeSerializedSeq && applied - 1 < seq) {
-    seq = applied - 1;
+  if (!_revisionTreeCanBeSerialized) {
+    return commitSeq;
   }
 
-  // now actually advance it if we can
-  if (seq > _revisionTreeSerializedSeq) {
-    _revisionTreeSerializedSeq = seq;
-  }
+  applyUpdates(commitSeq, lock);  // always apply updates...
 
-  return _revisionTreeSerializedSeq;
-}
+  // applyUpdates will make sure we will have a valid tree
+  TRI_ASSERT(_revisionTree != nullptr);
+  TRI_ASSERT(_revisionTree->depth() == revisionTreeDepth);
 
-rocksdb::SequenceNumber RocksDBMetaCollection::serializeRevisionTree(
-    std::string& output, rocksdb::SequenceNumber commitSeq, bool force) {
-  std::unique_lock<std::mutex> guard(_revisionTreeLock);
+  bool neverDone = _revisionTreeSerializedSeq == _revisionTreeCreationSeq;
+  bool coinFlip = RandomGenerator::interval(static_cast<uint32_t>(5)) == 0;
+  bool beenTooLong = 30 < std::chrono::duration_cast<std::chrono::seconds>(
+                              std::chrono::steady_clock::now() - _revisionTreeSerializedTime)
+                              .count();
+  TRI_IF_FAILURE("RocksDBMetaCollection::forceSerialization") {
+    coinFlip = true;
+  }
 
-  if (_logicalCollection.useSyncByRevision()) {
-    if (!_revisionTree && !haveBufferedOperations()) {  // empty collection
-      return commitSeq;
-    }
-    applyUpdates(commitSeq);  // always apply updates...
- 
-    // applyUpdates will make sure we will have a valid tree
-    TRI_ASSERT(_revisionTree != nullptr);
-    TRI_ASSERT(_revisionTree->depth() == revisionTreeDepth);
-
-    bool neverDone = _revisionTreeSerializedSeq == _revisionTreeCreationSeq;
-    bool coinFlip = RandomGenerator::interval(static_cast<uint32_t>(5)) == 0;
-    bool beenTooLong = 30 < std::chrono::duration_cast<std::chrono::seconds>(
-                                std::chrono::steady_clock::now() - _revisionTreeSerializedTime)
-                                .count();
-    TRI_IF_FAILURE("RocksDBMetaCollection::forceSerialization") {
-      coinFlip = true;
-    }
+  TRI_IF_FAILURE("RocksDBMetaCollection::serializeRevisionTree") {
+    return _revisionTreeSerializedSeq;
+  }
+  if (force || neverDone || coinFlip || beenTooLong) {  // ...but only write the tree out sometimes
+    _revisionTree->serializeBinary(output);
+    _revisionTreeSerializedSeq = commitSeq;
+    _revisionTreeSerializedTime = std::chrono::steady_clock::now();
 
-    TRI_IF_FAILURE("RocksDBMetaCollection::serializeRevisionTree") {
-      return _revisionTreeSerializedSeq;
-    }
-    if (force || neverDone || coinFlip || beenTooLong) {  // ...but only write the tree out sometimes
-      _revisionTree->serializeBinary(output);
-      _revisionTreeSerializedSeq = commitSeq;
-      _revisionTreeSerializedTime = std::chrono::steady_clock::now();
+    if (_revisionTreeSerializedSeq < _revisionTreeCreationSeq) {
+      _revisionTreeCreationSeq = _revisionTreeSerializedSeq;
```

Comment:
This is interesting. Originally, when a tree is first built, `_revisionTreeSerializedSeq` is set to be equal to `_revisionTreeCreationSeq`, and from then on, it always is >=. However, when we rebuild a tree after a failure, we set `_revisionTreeSerializedSeq` back to 0 (and delete the persisted tree).  That is, from then on the test in line 687 is indeed incorrect! We should change line 687 to test for `<=`. However, this does no big harm, since sooner or later, the tree is rebuilt, or some other condition kicks in (CoinFlip or beenTooLong).

Reply:
Done.